docs/versions/v26.4.5/guide/evolve/aspects/warnings.md

```
@@ -0,0 +1,37 @@
+
+## Aspect Invariants
+
+The following invariants define the fundamental behaviour of pipeline aspects:
+
+- Aspects do not change pipeline types or topology
+- Aspects may have side effects
+- Aspects may observe or persist data
+- Aspects are not allowed to alter pipeline control flow
+- Any aspect that does require data transformation must be modeled as a Step, not an Aspect
+
+These warnings are not validation errors and will not prevent compilation, but they may indicate potential configuration issues that users should review.
```

markdownlint-cli2 (0.22.1)

[warning] Files should end with a single newline character

(MD047, single-trailing-newline)